feat: support embellishments and non-punctuation delimiters

theseanl opened the pull request:
Previously, it couldn't attach arguments for macros whose argspec
 - involves embellishments `e{_}`, `E{^}{default}`,
 - or non-punctuation delimiters such as `r^_`

This PR adds support for it. The second use-case may be marginal, I initially tried to make embellishments work, then realized that the same code would enable the second feature and unifies some codes.

Since a single embellishment argspec can represent multiple arguments, such as in `e{^_}` where each token `^` and `_` would represents 2 arguments,
`gobbleSingleArgument`s signature was changed to return `Ast.Arguments | Ast.Arguments[] | null`.

Currently, this just puts embellishment tokens as `openMark` in Ast. Would it be preferable to have a new separate property for embellishments?

Next goal would be to support/test macro expansions for such arguments. Once it's done, I'd like to work on #24.

Please take a look and provide comments, and feel free to make your own changes!

siefkenj commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -212,3 +220,69 @@ function partialStringMatch(node: Ast.Node, token: Ast.Node): boolean {
+        return [openMarkPos, closeMarkPos];
+    }
+}
+function stringifyGroup(group: ArgSpec.Group): string {
```
I think the `printRaw` function already does this.

theseanl replied:
Thank you for point out. I believe its fixed now.

siefkenj commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -212,3 +220,69 @@ function partialStringMatch(node: Ast.Node, token: Ast.Node): boolean {
+    return Object.assign({}, node, { content });
+}
+
+// Find the position of the open brace and the closing brace.
```
Use `/** ... */` style comments for doc strings.

theseanl replied:
Fixed.

siefkenj commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -212,3 +220,69 @@ function partialStringMatch(node: Ast.Node, token: Ast.Node): boolean {
+// Find the position of the open brace and the closing brace.
+// Returns undefined if the brace isn't found.
+// This may mutate `nodes`.
+function findBracePositions(nodes: Ast.Node[], startPos: number,
```
This function looks to duplicate functionality already in `gobbleSingleArgument`

siefkenj replied:
Ah, I see you moved this function outside of the body.

siefkenj commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -188,6 +171,31 @@ export function gobbleSingleArgument(
+                );
+                return [];
+            });
+            for (let token of tokens) {
```
Use `const` if you're not mutating the value.

theseanl replied:
Fixed.

siefkenj commented on packages/unified-latex-ctan/package/minted/libs/argument-parser.ts:
```diff
@@ -16,17 +16,17 @@ const argSpecRDelim: { [delim: string]: Node } = {};
     if (match.group(nextNode)) {
         const mandatoryArg = gobbleSingleArgument(nodes, argSpecM, startPos);
-        codeArg = mandatoryArg.argument;
+        codeArg = mandatoryArg.argument as Argument;
```
Prefer testing with `Array.isArray` and throwing an error over casting the type.

theseanl replied:
Fixed at [201461f](https://github.com/siefkenj/unified-latex/pull/42/commits/201461fb3a721645878877726f17d44fc9e396e8)

siefkenj commented:
Thanks for the contribution!

It looks like you have some tests failing...

Here are some thoughts:
 1. It would be really nice not to change the signature of `gobbleSingleArgument`, but I cannot think of a good way around it. One possibility is to change it to always return `[] | null`
 2. I think more of the existing infrastructure can be reused. The algorithm for `e{x}` should be very similar to `tx m`. I think you can reuse the algorithms for grabbing those arguments. A possible idea for `e{xyz}` is that you keep a record of whether `x,y,z` has been captured. If one of them has been, you can re-call the function with `y,z` (or whatever hasn't yet been captured). This could potentially result in a way to keep `gobbleSingleArgument` returning a singleton. If it always returned the first argument it could gobble, then `gobbleArgument` would know that on a `type=="embellishment"` argument, it would need to call until nothing was gobbled. Then then it could rearrange the argument list to appear in the order specified by the oringinal macro spec.
 3. It would be nice to keep `e{^}` consistent with the parsing of `^`. Currently `^` parses as a macro with `escapeToken == "^"`. I think it's reasonable to parse `e{x}` as 
```
{
                    "type": "argument",
                    "content": [
              {
                    "type": "macro",
                    "content": "^",
                    "escapeToken": "",
                    "args": [
                        {
                            "type": "argument",
                            "content": [
                                {
                                    "type": "string",
                                    "content": "2"
                                }
                            ],
                            "openMark": "{",
                            "closeMark": "}"
                        }
                    ]
                }
                ]
                openMark: "",
                closeMark: ""
```

That solution is not exactly elegant, though...

theseanl commented:
# Regarding re-using existing algorithm for `t`

I wasn't aware that `optionalToken` shares similar traits. However, the way I see it is the other way around: Parsing `optionalToken` should use the new infrastructure `findBracePositions`.
A convoluted example, but the current algorithm for `t` can't correctly figure out arguments for the following:
```tex
\NewDocumentCommand{\asdf}{ t^ r__ }{\IfBooleanTF{#1}{123}{456} #2}
\asdf^_789_
```
It should be expanded to `123 789`. 
I will fix this and include it as a test case.

# Signature of `gobbleSingleArgument`

One idea I have thought was to artificially decompose `e{^-_}` into `e{^} e{-} e{_}`. However it seemed that signatures are intimately tied to the raw `.tex` source and ArgSpec is assumed to be a faithful representation of it, there is even a warning https://github.com/siefkenj/unified-latex/blob/f2b8c107aed4c7ae585ebbd832ef9085cee35a46/packages/unified-latex-util-argspec/README.md?plain=1#L30

> A possible idea for `e{xyz}` is that you keep a record of whether `x,y,z` has been captured. If one of them has been, you can re-call the function with `y,z` (or whatever hasn't yet been captured). This could potentially result in a way to keep `gobbleSingleArgument` returning a singleton

It seems doable, but then the caller of `gobbleSingleArgument` should maintain some kind of state, about the embellishment tokens we have matched, its index and so on. Such a logic would then go inside the body of `gobbleArguments`, but it doesn't look like the right place for that logic..

If you prefer to have its return type `Ast.Arguments[] | null`, then I'll refactor that.

siefkenj commented:
I think the real issue with changing `e{xy}` to `e{x} e{y}` is that, I believe that `e{xy}` allows `x`/`y` to occur in either order, where `e{x} e{y}` would not.

The mutation warning is just that: a warning against mutating. However, if you create a new `ArgSpec` or make a deep copy and mutate the deep copy, it will be fine.

I'm leaning more towards making `gobbleSingleArgument` no return an array and moving some looping logic into `gobbleArgument`.

theseanl commented:
> I think the real issue with changing e{xy} to e{x} e{y} is that, I believe that e{xy} allows x/y to occur in either order, where e{x} e{y} would not.

I wasn't aware of that! Thank you for pointing it out. I've updated code to implement this behavior.
Now given with this order-agnostic behavior of embellishment, moving the looping logic to `gobbleArguments` feels slightly more natural. However it's still annoying to pass around matched embellishment token and its index between `gobbleArguments` and `gobbleSingleArgument`, so I kept the signature in the end.

Also I've added a test for `\asdf^_789_` as promised.

siefkenj commented:
FYI, I just merged #45 which switches the tests to vitest. They might run successfully on your computer now :-)

theseanl commented:
`until` also manifests the same issue as `optionalToken`, so I went on to fix that as well :)
If you look at the code, it should be obvious where I am going. We will be able to handle another edge case: #46.
How does it look? What's your thought on `gobbleSingleArgument`? Also could you elaborate more on `escapeToken == "^"` (if this is still an issue)?

siefkenj commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -212,3 +210,97 @@ function partialStringMatch(node: Ast.Node, token: Ast.Node): boolean {
+    let openMarkPos = startPos;
+    let closeMarkPos: number | null = startPos;
+    if (openMark) {
+        if (!match.anyString(currNode)) return;
```
Is `whitespace` forbidden during a brace search? Not always, right?

theseanl replied:
As I understand, `findBracePositions` is always (or should be) called after `gobbleWhitespaces` is called.
So `currNode` is always a non-whitespace node. That said, I realized that I should call `gobbleWhitespaces` between embellishment loops, fixed in https://github.com/siefkenj/unified-latex/pull/42/commits/052b4ff7c4ffbe8017ff0a780054133f6b77bf87
But this may become obsolete once the gobbleSingleArgument function is refactored to return single argument.

siefkenj commented on packages/unified-latex-util-arguments/tests/attach-arguments-in-array.test.ts:
```diff
@@ -197,5 +197,120 @@ describe("unified-latex-util-arguments", () => {
+        ]);
+
+        // embellishments
+        nodes = strToNodes("\\xxx^a\\xxx_b\\xxx_b^a");
```
Can you add another test with `\xxx^{a_b}`?

theseanl replied:
Done in https://github.com/siefkenj/unified-latex/pull/42/commits/5839b7364cbe623781d11f501a7bf1ccfd3c4db1

siefkenj commented:
Thanks for this! I left a few comments. Also, please run Prettier on all modified files.

I'd still like to move some of the complexity to `gobbleArguments` and have `gobbleSingleArugment` only return one thing rather than an array.

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -34,15 +34,15 @@ export function gobbleSingleArgument(
     const gobbleWhitespace = (argSpec as ArgSpec.LeadingWhitespace)
         .noLeadingWhitespace
-        ? () => {}
+        ? () => { }
```
I believe my vscode prettier extension picked up a different config. Fixed in https://github.com/siefkenj/unified-latex/pull/42/commits/ee97bad5c1b7f9e05eeb1ff0e5c2954a9a6459a3

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -212,3 +210,97 @@ function partialStringMatch(node: Ast.Node, token: Ast.Node): boolean {
+    if (Array.isArray(arg)) {
+        throw new Error(`Expected an argspec to gobble single argument, but got ${arg.length}`);
+    }
+}
```
Done in https://github.com/siefkenj/unified-latex/pull/42/commits/ee97bad5c1b7f9e05eeb1ff0e5c2954a9a6459a3

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -157,37 +138,54 @@ export function gobbleSingleArgument(
+            do { // Try finding match until there is no more
+                hasMatch = false;
+                for (let i = 0; i < tokens.length; i++) {
+                    if (argument[i]) continue;
```
Done in https://github.com/siefkenj/unified-latex/pull/42/commits/5897e7d20dfc3e40639b07d4029096b5ef38b573

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -157,37 +138,54 @@ export function gobbleSingleArgument(
+                    const bracePos = findBracePositions(nodes, currPos, token);
+                    if (!bracePos) continue;
+                    let argNode = nodes[bracePos[0] + 1];
+                    argument[i] = arg(match.group(argNode) ? argNode.content : argNode, {
```
Yes, and that's being implemented here.

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -157,37 +138,54 @@ export function gobbleSingleArgument(
+                    if (argument[i]) continue;
+                    const token = tokens[i];
+                    const bracePos = findBracePositions(nodes, currPos, token);
+                    if (!bracePos) continue;
```
`bracePos`'s type is `[number,number] | undefined`, so yes.

theseanl commented:
I've added eslint and added `curly` rule which enforces statements in if ...  else to be enclosed in a separate block.
I couldn't find a prettier config, and I doubt it's being consistently used in codes in this repository, for example prettier seems to have some peculiar rules for trailing commas and it is not consistently enforced.
I've tried to add one that matches the overall coding style of the repository. It revealed some obvious "error"s (in the sense of prettier's) whose fix was trivial.

siefkenj commented on .prettierrc.json:
```diff
@@ -0,0 +1,4 @@
+{
```
I think this can go in a `"prettier": ...` section in `package.json`, if it's not already there.

theseanl replied:
Done in https://github.com/siefkenj/unified-latex/pull/42/commits/574d40963a3a57d91f1018f5dcb6abc0eceb2212

siefkenj commented on .eslintrc.js:
```diff
@@ -13,5 +16,14 @@ module.exports = {
+        "no-case-declarations": "off",
+        "prefer-const": "off",
+        "no-empty": "off",
+        "no-var": "off",
```
This one should be okay. I don't think `var` is used anywhere...

theseanl replied:
There were some at `print-raw.ts`. I've removed the rule and changed occurrences of `var` to `let` in https://github.com/siefkenj/unified-latex/pull/42/commits/0ac4a930a24b8ee7f5530af36e2af2aa6964b01f.

siefkenj commented on packages/structured-clone/index.ts:
```diff
@@ -1,8 +1,8 @@
 // globalThis polyfill from https://mathiasbynens.be/notes/globalthis
 (function () {
-    if (typeof globalThis === "object") return;
+    if (typeof globalThis === "object") {
```
Some day `structuredClone` will have wide support and I can finally drop this nonsense :-/

theseanl replied:
According to https://developer.mozilla.org/en-US/docs/Web/API/structuredClone, it is supported in nodejs \ge 17.0.0, but not in nodejs workers. I guess this nodejs workers are the most concerning part?

siefkenj replied:
I'm also concerned about browser support :-)

theseanl replied:
Other browsers look fine but Safari support seems problematic. I just tried running tests with `delete globalThis.structuredClone` before every tests and it passed all (except for those failing before), maybe having this kind of automated testing would be good.

siefkenj commented:
Thanks for the linter updates! Let me know when this is ready for another review :-)

theseanl commented:
Refactored `gobbleSingleArgument`. I don't really like the additional complexity introduced by artificially separating logic into two places, but I guess I can live with it. Now `gobbleArguments` always create a `structuredClone` of the `argSpec` because `gobbleSingleArgument` may mutate it. It may as well check for `spec.type === 'embellishment'` and create a clone only in that case, but I wanted to make `gobbleArgument` agnostic to individual argspec types.

theseanl commented:
Also I think the current tsconfig is somewhat faulty,
```
cd packages/unified-latex
npm run build
```
fails with TS errors which shouldn't be an error, and `describe` and `expect` isn't resolved in vscode.

siefkenj commented:
> Also I think the current tsconfig is somewhat faulty,
> 
> ```
> cd packages/unified-latex
> npm run build
> ```
> 
> fails with TS errors which shouldn't be an error, and `describe` and `expect` isn't resolved in vscode.

Did you do a general build first? If you haven't built the dependencies, it will fail.

I have `vitest/global` listed in `tsconfig.build.json`. That is supposed to explain to typescript that `describe` and `it`, etc.. all exist. I don't know why it doesn't work...

theseanl commented:
Turned out that doing a general build fixes the issue with `npm run build`. It initially failed, so I fixed some type errors in structured-clone that were causing the issue.

Regarding tsconfig, `tsconfig.build.json` explicitly `excludes` any test files, and any other tsconfig inherits from it, so IDE think test files are not part of the project and doesn't see `vitest/global` defined in tsconfig while trying to resolve types.

IMO, `tsconfig.build.json` should inherit from `tsconfig.json`, and excluding test files should be done only in `tsconfig.build.json` file. Each package directory should contain two `tsconfig.build.json` and `tsconfig.json`, and `npm run build` command should invoke `tsc` with `tsc -b tsconfig.build.json`. I'm not confident that I'll be able to change every relevant part, so it'd be nice if you could fix it :)

siefkenj reviewed (commented):
This is looking really close!

siefkenj commented:
> IMO, `tsconfig.build.json` should inherit from `tsconfig.json`, and excluding test files should be done only in `tsconfig.build.json` file. Each package directory should contain two `tsconfig.build.json` and `tsconfig.json`, and `npm run build` command should invoke `tsc` with `tsc -b tsconfig.build.json`. I'm not confident that I'll be able to change every relevant part, so it'd be nice if you could fix it :)

Could you do a separate PR that shows what you're talking about in a couple of the workspaces? The `tsconfig` business is really a pain and I've been trying to figure out how to get it right for a long time...

Alternatively, importing `describe`, `it` and friends in each test file will get rid of the typescript warnings.

theseanl commented on packages/unified-latex-util-argspec/libs/argspec-types.ts:
```diff
@@ -34,14 +34,14 @@ interface OptionalToken extends LeadingWhitespace, AstNode {
+export interface Embellishment extends DefaultArgument, AstNode {
     type: "embellishment";
-    embellishmentTokens: string[];
+    embellishmentTokens: (Group | string)[];
```
Group here is for $e{{x}{y}}$. I agree with you that this should be fixed at pegjs level alongside with https://github.com/siefkenj/unified-latex/issues/46. But working on pegjs would broaden the scope of this pr too much, I'd prefer to fix it in another PR.

siefkenj replied:
I just merged #49 which should mean that groups are no longer returned as `embellishmentTokens`.

theseanl commented on packages/unified-latex-util-arguments/libs/gobble-single-argument.ts:
```diff
@@ -157,37 +170,60 @@ export function gobbleSingleArgument(
+            const tokens = normalizeEmbellishmentTokens(
+                argSpec.embellishmentTokens
+            );
+            argSpec.embellishmentTokens = tokens; // ArgSpec is mutated here
```
So for example, are you OK with mutating it in `gobbleArguments` function? What exactly is the goal you want by not mutating it here?

siefkenj commented:
I would prefer to recreate it on each call rather than mutate it.

On September 17, 2023 10:12:05 AM PDT, Sean Lee ***@***.***> wrote:
>@theseanl commented on this pull request.
>
>
>
>>              if (currPos < nodes.length) {
>                 currPos++;
>             }
>             break;
>         }
>+        case "embellishment": {
>+            // Split tokens into single characters
>+            const tokens = normalizeEmbellishmentTokens(
>+                argSpec.embellishmentTokens
>+            );
>+            argSpec.embellishmentTokens = tokens; // ArgSpec is mutated here
>
>So for example, are you OK with mutating it in `gobbleArguments` function? What exactly is the goal you want by not mutating it here?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/siefkenj/unified-latex/pull/42#discussion_r1328119277
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

siefkenj commented:
Agreed. Can you just make it throw an error for now if it encounters a group and we can fix it in another PR

On September 17, 2023 10:10:19 AM PDT, Sean Lee ***@***.***> wrote:
>@theseanl commented on this pull request.
>
>
>
>>      type: "embellishment";
>-    embellishmentTokens: string[];
>+    embellishmentTokens: (Group | string)[];
>
>Group here is for $e{{x}{y}}$. I agree with you that this should be fixed at pegjs level alongside with https://github.com/siefkenj/unified-latex/issues/46. But working on pegjs would broaden the scope of this pr too much, I'd prefer to fix it in another PR.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/siefkenj/unified-latex/pull/42#discussion_r1328119056
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

siefkenj commented:
@theseanl Were you planning on finishing the last changes?

theseanl commented:
I need to eventually get it done, because it is needed for my other works, but I've been rather busy and less motivated to refactor an already working code to meet a goalpost that keeps moving. Conceptual distinction of singleton and multiple objects feel artificial to me, I even think that such a tendency may be influenced by one's linguistic backgrounds.. I'd be ok if you take over and address the points you've mentioned in reviews. I'll try to find some time within a few weeks.

siefkenj commented:
@theseanl Thanks for your hard work! I have fixed the last of the issues. I couldn't figure out how to get pushing to your branch working, so I created PR #57 You can either pull from that branch, or I can close this PR and merge #57 instead.

theseanl commented:
Thank you for picking up the slack!
